src/DataStructures/Tensor/EagerMath/Norms.hpp

```diff
@@ -0,0 +1,141 @@
+  }
+  return pointwise_l2_normsq;
+}
+}  // namespace L2Norm_detail
+
+/*!
+ * \ingroup TensorGroup
+ * \brief Compute point-wise Frobenius norm of arbitrary tensors.
+ *
+ * \details
+ * At each point \f$ p\f$ in the element, this function computes the
+ * point-wise Frobenius norm of a given tensor of arbitrary rank. If the tensor
+ * \f$A\f$ has rank \f$n\f$ and dimensionality \f$D\f$, then its Frobenius norm
+ * at point \f$ p\f$ is computed as: \f{equation}
+ * ||A||_2^p =
+ *    \left(\sum^{D-1}_{i_1=0}\sum^{D-1}_{i_2=0}\cdots \sum^{D-1}_{i_n=0}
+ *          |A^p_{i_1 i_2 \cdots i_n}|^2 \right)^{1/2},
+ * \f}
+ * where both contra-variant and co-variant indices are shown as lower indices.
+ */
+template <typename DataType, typename Symm, typename IndexList>
+Scalar<DataType> pointwise_l2_norm(
+    const Tensor<DataType, Symm, IndexList>& tensor) noexcept {
+  return Scalar<DataType>{
+      sqrt(get(L2Norm_detail::pointwise_l2_norm_square(tensor)))};
+}
+
+/*!
+ * \ingroup TensorGroup
+ * \brief Compute Frobenius norm of arbitrary tensors reduced over an element.
+ *
+ * \details
+ * Computes the RMS value of the point-wise Frobenius norm of a given tensor
+ * (of arbitrary rank) over all points in an element. If the tensor \f$A\f$ has
+ * rank \f$n\f$ and dimensionality \f$D\f$, and the element has \f$N\f$ points,
+ * then its element-reduced Frobenius norm is computed as:
+ * \f{equation}
+ *  ||A||_2 =
+ *     \left(\frac{1}{N}\sum^{N-1}_{p=0}
+ *          \left(\sum^{D-1}_{i_1=0}\sum^{D-1}_{i_2=0}\cdots
+ *                \sum^{D-1}_{i_n=0} |A^p_{i_1 i_2 \cdots i_n}|^2 \right)
+ * \right)^{1/2},
+ * \f}
+ * where both contra-variant and co-variant indices are shown as lower indices,
+ * and the superscript \f$p\f$ indexes points in the element.
+ *
+ * \warning This function reduces the Frobenius norm over the element, not the
+ * whole domain.
+ */
+template <typename DataType, typename Symm, typename IndexList>
+double l2_norm(const Tensor<DataType, Symm, IndexList>& tensor) noexcept {
+  const auto pointwise_l2_normsq =
+      L2Norm_detail::pointwise_l2_norm_square(tensor);
+  using Plus = funcl::Plus<funcl::Identity>;
+  return sqrt(alg::accumulate(get(pointwise_l2_normsq), 0., Plus{}) /
+              tensor.begin()->size());
+}
+
+namespace Tags {
+/*!
+ * \ingroup DataBoxTagsGroup
+ * \ingroup DataStructuresGroup
+ * Point-wise Frobenius norm of a tensor. See pointwise_l2_norm() for details.
+ */
+template <typename Tag>
+struct PointwiseL2Norm : db::SimpleTag {
+  using type = Scalar<typename Tag::type::type>;
+  static std::string name() noexcept {
+    return "PointwiseL2Norm(" + db::tag_name<Tag>() + ")";
+  }
+};
+/*!
+ * \ingroup DataBoxTagsGroup
+ * \ingroup DataStructuresGroup
+ * Computes the point-wise Frobenius norm of a tensor. See pointwise_l2_norm()
+ * for details.
+ */
+template <typename Tag>
+struct PointwiseL2NormCompute : PointwiseL2Norm<Tag>, db::ComputeTag {
```

Comment:
Since we‘ll be computing norms of sets of tensors, I think a version of these tags that works on `Variables` would be useful. This depends on how you‘ll use it Prayush. If you add a whole list of these compute tags to the DataBox where you could add one for a Variables, then please add a Variables overload for this tag. That could also be optimized (perhaps later) to work with the contiguous memory layout of the Variables.

 E.g. `Tags::Flux` in Prefixes.hpp gives an example how to write the Tensor/Variables overload.

Reply:
okay, I've added an overload for Variables. (For my use case of observing constraint norms, I *won't* have all constraints available inside a Variables container as they'll each be individually added to the databox by their respective compute tags.)